cluster-controller/src/main/java/io/strimzi/controller/cluster/resources/TopicController.java

```diff
@@ -173,33 +185,18 @@ public static TopicController fromConfigMap(ConfigMap kafkaClusterCm) {
 
         String config = kafkaClusterCm.getData().get(KEY_CONFIG);
         if (config != null) {
-            topicController = new TopicController(kafkaClusterCm.getMetadata().getNamespace(),
+            String namespace = kafkaClusterCm.getMetadata().getNamespace();
+            topicController = new TopicController(namespace,
                     kafkaClusterCm.getMetadata().getName(),
                     Labels.fromResource(kafkaClusterCm));
 
             JsonObject json = new JsonObject(config);
 
-            String image = json.getString(TopicController.IMAGE_FIELD);
-            if (image != null) {
-                topicController.setImage(image);
-            }
-
-            String topicNamespace = json.getString(TopicController.NAMESPACE_FIELD);
-            if (topicNamespace != null) {
-                topicController.setTopicNamespace(topicNamespace);
-            }
-
-            String reconciliationIntervalMs = json.getString(TopicController.RECONCILIATION_INTERVAL_FIELD_MS);
-            if (reconciliationIntervalMs != null) {
-                // TODO : add parsing and validation
-                topicController.setReconciliationIntervalMs(reconciliationIntervalMs);
-            }
-
-            String zookeeperSessionTimeoutMs = json.getString(TopicController.ZOOKEEPER_SESSION_TIMEOUT_FIELD_MS);
-            if (zookeeperSessionTimeoutMs != null) {
-                // TODO : add parsing and validation
-                topicController.setZookeeperSessionTimeoutMs(zookeeperSessionTimeoutMs);
-            }
+            topicController.setImage(json.getString(TopicController.IMAGE_FIELD, DEFAULT_IMAGE));
+            topicController.setTopicNamespace(json.getString(TopicController.NAMESPACE_FIELD, namespace));
```

Comment:
It seems a bit mixed up to me that we get the `namespace` from the CM, pass that to the `TopicController` ctor, and then here we lookup the `NAMESPACE_FIELD` from the `json`, defaulting it to `namespace`, so we can override the namespace we passed to the ctor in the first place! Shouldn't we just look it up from the `json` in the first place, and then pass that to the `TopicController` ctor?

Reply:
the `namespace` in the JSON is not mandatory. If you omit this, the TC will watch for topic CM created in the same namespace as the deployment. Using the `namespace` field in the JSON configuration we allow to the TC to watch for topic CM in a different namespace where the TC is running (i.e. a namespace where a user application is running).

Reply:
Not sure if I didn't explain it well, but what I meant was to change ~line 188 to:

```
JsonObject json = new JsonObject(config);
String namespace = kafkaClusterCm.getMetadata().getNamespace();
topicController = new TopicController(json.getString(TopicController.NAMESPACE_FIELD, namespace),
                     kafkaClusterCm.getMetadata().getName(),
                     Labels.fromResource(kafkaClusterCm));
```

It's not a big deal, it just seems strange that you factored out the local `namespace` to pass to the ctor, only to be possibly overriding it here, when you could just pass the correct value to the ctor initially.

Reply:
The `namespace` parameter for the `TopicController` ctor is really the namespace where the TC is deployed (as you can see it passes that parameter to the super ctor of the `AbstractCluster`).
Inside the `TopicController` ctor happens that the field `topicNamespace` (where the TC should watch for topic CM) is set exactly to that namespace as default.

If I pass the namespace from the JSON as you are proposing it seems that such parameter is the namespace where the TC should watch for the topic CM ... but it's not.
(In fact you can notice that we call `setTopicNamespace` on the TC instance not `setNamespace`.)
I hope I explained well here :-(

Reply:
OK, I understand. There's scope for confusion here. I wonder whether defaulting the `this.topicNamespace = namespace` in the `TopicController` ctor is really a good idea then. Better to have two completely different namespace things in the `TopicController` itself and let clients set them to the same thing if they need to, than to say "hey there's two namespacey things going on, and by default I'm going to conflate them". That's just confusing!

Reply:
Well we have two separate things which are `namespace` and `topicNamespace` with their different meanings.
Because `TopicController' is a cluster resource (as `KafkaCluster` and so on but remember that we should change the naming here), it has the same ctor parameters as the other and the first one is the `namespace` it lives.
The "other" namespace is a configuration parameter for its behaviour not related on where it lives. I don't see a confusion. From a user point of view we also explain it in the documentation :

```
`namespace`:: The Kubernetes namespace (OpenShift project) in which the topic controller watches for topic ConfigMaps.
Default is the namespace where the topic controller is running
```

Just one nit, we could call this JSON field `topicNamespace` explicitly and not just `namespace` ?

Reply:
>  I don't see a confusion.

The confusion is in `TopicController.<init>` where it somewhat arbitrarily decides to default the topic namespace to the namespace where the topic controller is running. If they're really separate things then there's no good reason to default the one to the other like that. Sure, let the caller of `TopicController` set them to the same thing if they need to be the same, but don't do that by default.

Reply:
> Just one nit, we could call this JSON field `topicNamespace` explicitly and not just `namespace` ?

A topic cannot have a namespace, so that might be confusing. Would something like `managedNamespace`, `watchedNamespace`, `monitoredNamespace` be better?

Reply:
> Sure, let the caller of TopicController set them to the same thing if they need to be the same, but don't do that by default.

I don't agree on that. I don't see a big deal if something like that is well documented. The first idea was just having topic controller watching topic CMs where it was deployed but then we saw a limitation on that and the user could have a different namespace where topic CMs are deployed together with the application that will use the Kafka cluster.

Regarding the name proposals I'm fine with `watchedNamespace` to have a more consistent naming with the fact that we are using "watchers" for doing that.

Reply:
> I don't agree on that. 

What I'm proposing wouldn't change that behaviour, as observed by someone deploying the TC. It would simply make it clearer in the `TopicController` code that there are two distinct things.